### PR TITLE
Add Dockerfile + concept of a DATA_DIR

### PR DESCRIPTION
**Pull request by cooperlees**
- Add a Dockerfile so people can run this bot in a docker container
- Stuck with recommendation of running with python3.9 for now
      - Will later test with 3.11 + supply fixes if I get this working ...
- Added a DATA_DIR env param to use to choose the directory to write data we want persistent across docker container restarts to be written to
      - We default to CWD like the code does today - we just explicitly pass it to functions / classes

Test:
- `docker build -t gpt3discord .`
```
crl-m1:GPT3Discord cooper$ docker image ls
REPOSITORY                     TAG               IMAGE ID       CREATED          SIZE
gpt3discord                    latest            6d2832af2450   69 seconds ago   356MB
```

- `docker run --name gpt3discord -v ~/gpt3discord/env:/bin/.env -v ~/gpt3discord/data:/data gpt3discord`
```
crl-m1:GPT3Discord cooper$ docker run --name gpt3discord -v ~/gpt3discord/env:/bin/.env -v ~/gpt3discord/data:/data gpt3discord
None of PyTorch, TensorFlow >= 2.0, or Flax have been found. Models won't be available and only tokenizers, configuration and file/data utilities can be used.
Downloading: 100%|██████████| 1.04M/1.04M [00:00<00:00, 1.79MB/s]
Downloading: 100%|██████████| 456k/456k [00:00<00:00, 921kB/s]
Downloading: 100%|██████████| 1.36M/1.36M [00:00<00:00, 2.23MB/s]
Downloading: 100%|██████████| 665/665 [00:00<00:00, 179kB/s]
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/cogs/image_prompt_optimizer.py", line 40, in __init__
    with image_pretext_path.open("r") as file:
  File "/usr/local/lib/python3.9/pathlib.py", line 1252, in open
    return io.open(self, mode, buffering, encoding, errors, newline,
  File "/usr/local/lib/python3.9/pathlib.py", line 1120, in _opener
    return self._accessor.open(self, flags, mode)
FileNotFoundError: [Errno 2] No such file or directory: '/data/image_optimizer_pretext.txt'
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
```

See it work in a discord server. Do we want to fix any of these warnings?

### CODE REVIEW

**Comment by cooperlees** [comment]
I will read the code, but if you know, can you save me some time please.

Where does this bot save it's state on the filesystem? I ask cause I'll need to bind mount a persistent directory for it to write to so when we update the docker container the state is maintained ...

**Comment by Kav-K** [comment]
> 

Hi Cooper, first of all, thank you SO so much for the initiative, I really appreciate it.

The bot saves state a few different ways on the file system. It saves usage data in a `usage.txt` in the CWD. It saves combined images when image generations are happening as `combined.png` in the CWD. It saves individual generated dall-e images in a `dalleimages` directory in the CWD. Finally, tempfiles are used, with the `tempfile` library in python, saved in ` \TMP` in a unix instance

**Comment by cooperlees** [comment]
Ok - I tried to find and change all the CWD usage that looked like it would be nice to save on docker container restarts. Please point out any I have missed. I've been playing with the bot all afternoon on a discord server. Thanks for your work there!

**Comment by Kav-K** [comment]
Thank you very much Cooper, will take a look soon!

**Comment by Kav-K** [comment]
Hey Cooper, the changes you've made look great. I just wanted to run a few things by you:
- No datadir and changes are needed for the writing of the `bot.pid` file in main.py, because this is just a static file we don't care about saving between restarts, right?
- The bot creates image files in a `dalleimages/` directory, and in `/tmp`, do we need to do anything special to handle these cases in Docker?

**Comment by cooperlees** [comment]
> * No datadir and changes are needed for the writing of the `bot.pid` file in main.py, because this is just a static file we don't care about saving between restarts, right?

Correct, if the processes crashes or we restart / stop and start, docker restarts the whole container. So the PID file will never exist at start. Technically, we don't even really need a PID file at all in docker, it's overhead. Neither does systemd or modern ways to run daemons. How do you use the PID file out of interest?

> * The bot creates image files with a TTL of 1 hour in a `dalleimages/` directory, and temporary named files in `/tmp`, do we need to do anything special to handle these cases in Docker?

Because of the TTL and nature of putting files in /tmp, I felt it didn't need to be persistent across restarts. These all felt like caching to me. Due to that, I don't think it needed to be persisted across docker restart + upgrades. But please let me know if I've missed something.

If I am correct about the dalleimages dir, I'd argue it should move to being a directory in /tmp (or location that tempfile module chooses for each respective platform) to echo that nature.

Please do let me know thow if I have missed something that should be persistent ...

**Comment by Kav-K** [comment]
> > * No datadir and changes are needed for the writing of the `bot.pid` file in main.py, because this is just a static file we don't care about saving between restarts, right?
> 
> Correct, if the processes crashes or we restart / stop and start, docker restarts the whole container. So the PID file will never exist at start. Technically, we don't even really need a PID file at all in docker, it's overhead. Neither does systemd or modern ways to run daemons. How do you use the PID file out of interest?
> 
> > * The bot creates image files with a TTL of 1 hour in a `dalleimages/` directory, and temporary named files in `/tmp`, do we need to do anything special to handle these cases in Docker?
> 
> Because of the TTL and nature of putting files in /tmp, I felt it didn't need to be persistent across restarts. These all felt like caching to me. Due to that, I don't think it needed to be persisted across docker restart + upgrades. But please let me know if I've missed something.
> 
> If I am correct about the dalleimages dir, I'd argue it should move to being a directory in /tmp (or location that tempfile module chooses for each respective platform) to echo that nature.
> 
> Please do let me know thow if I have missed something that should be persistent ...

Re: the PID file, I'm very traditional, the only reason I have a PID file is to ensure multiple instances of the bot don't start, and so I can easily access the PID through github actions in order to kill an already running instance. My current github actions pipeline is very rudimentary, it simply does black formatting and then uploads the changes to my personal instance, there's no testing or checks done apart from black.

In terms of the `dalleimages` and `/tmp`, you're right and I think we're all good inthat regard, and I'll get rid of `dalleimages` folder in the future and just have all the images be in tmp. I'm not sure why I even decided to separate these two in the first place...

I'll be merging shortly after one more pass through to check for persistent files

**Comment by Kav-K** [comment]
@cooperlees Another question that I have actually:

In the upcoming future, I'm going to be adding pineconedb integration to this bot, to act as a vector database. Will this cause any problems with Docker? Pineconedb is interfaced with a package in python called https://pypi.org/project/pinecone-client/, I believe all communications for the client happen over HTTPS.

**Comment by cooperlees** [comment]
> @cooperlees Another question that I have actually:
> 
> In the upcoming future, I'm going to be adding pineconedb integration to this bot, to act as a vector database. Will this cause any problems with Docker? Pineconedb is interfaced with a package in python called https://pypi.org/project/pinecone-client/, I believe all communications for the client happen over HTTPS.

What benefit/functionality does this give? Does it cache questions and answers indefinitely? I’m no vector DB guru.

The main way to think about it is every re-build/upgrade of a docker container is a fresh start unless we put it in the DATA_DIR as
It’s a directory on the base operating system and is persistent. I also backup my persistent dirs so what ever is handy for recovery (e.g. the env file) should be there.

If there is benefit of keeping the vector DB state, it should go in the data dir. If it’s a nice to have cache or short lived it should go in /tmp / container file system and will be lost on the next rebuild / upgrade.

**Comment by Kav-K** [comment]
> > @cooperlees Another question that I have actually:
> > In the upcoming future, I'm going to be adding pineconedb integration to this bot, to act as a vector database. Will this cause any problems with Docker? Pineconedb is interfaced with a package in python called https://pypi.org/project/pinecone-client/, I believe all communications for the client happen over HTTPS.
> 
> What benefit/functionality does this give? Does it cache questions and answers indefinitely? I’m no vector DB guru.
> 
> The main way to think about it is every re-build/upgrade of a docker container is a fresh start unless we put it in the DATA_DIR as It’s a directory on the base operating system and is persistent. I also backup my persistent dirs so what ever is handy for recovery (e.g. the env file) should be there.
> 
> If there is benefit of keeping the vector DB state, it should go in the data dir. If it’s a nice to have cache or short lived it should go in /tmp / container file system and will be lost on the next rebuild / upgrade.

Got it, I understand now. 

The vectorDB will store embeddings to enable long term accurate conversation memory, the DB is not local and is hosted on a GCP instance provided by pinecone, the library interfaces with that instance with HTTPS, so I guess there is no need for any persistence here and we don’t have to worry about it in Docker. Does my logic follow?

**Comment by cooperlees** [comment]
Yup. If it’s a cloud service well all call the sockets will just do their things and communicate away …

If we need different API tokens or something I would imagine that goes into your env file config …

**Comment by Kav-K** [comment]
> Yup. If it’s a cloud service well all call the sockets will just do their things and communicate away …
> 
> If we need different API tokens or something I would imagine that goes into your env file config …

Yeah, it’ll just be the addition of another api token for pinecone and a few more connection params.

Alrighty I’m good to merge thank you so much for your work and responsiveness!

File: cogs/image_prompt_optimizer.py [hunk 1]
```diff
@@ -34,11 +34,12 @@ def __init__(
         self.deletion_queue = deletion_queue
 
         try:
+            image_pretext_path = self.converser_cog.data_path / "image_optimizer_pretext.txt"
```
Comment (Kav-K) [inline cogs/image_prompt_optimizer.py:37]: Neat, didn’t know you could do this

**Review by Kav-K — APPROVED** [review]

File: cogs/image_prompt_optimizer.py (hunk 1 above)
Comment (cooperlees) [reply cogs/image_prompt_optimizer.py:37]: I love pathlib. Remove os.path from your life today!

Comment (Kav-K) [reply cogs/image_prompt_optimizer.py:37]: Consider it done! :)

**Pull request merged**